### DISCUSSION TITLE
macos cross compile

### DISCUSSION THREAD

**Discussion by cstkingkey** [fltk-rs/fltk-rs#1142] [Q&A]
using https://github.com/tpoechtrager/osxcross

using cc wrapper "o64-clang-ld"
#!/bin/sh
/usr/local/osxcross/bin/x86_64-apple-darwin18-clang -fuse-ld=/usr/local/osxcross/bin/x86_64-apple-darwin18-ld $@

add dst.define("CMAKE_SYSTEM_NAME", "Darwin"); to fltk-sys/build/source.rs

running  CC="o64-clang-ld" CXX="o64-clang++-ld" LINK="o64-clang++-ld" cargo build --target x86_64-apple-darwin

end up with
 [ 93%] Linking CXX executable ../bin/fluid.app/Contents/MacOS/fluid

  --- stderr

  gmake: warning: -jN forced in submake: disabling jobserver mode.
  ld: warning: ignoring file /usr/lib64/libdl.so, building for macOS-x86_64 but attempting to link with file built for unknown-unsupported file format ( 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 )
  ld: warning: ignoring file ../lib/libfltk_images.a, building for macOS-x86_64 but attempting to link with file built for macOS-x86_64
  ld: warning: ignoring file ../lib/libfltk.a, building for macOS-x86_64 but attempting to link with file built for unknown-unsupported file format ( 0x21 0x3C 0x61 0x72 0x63 0x68 0x3E 0x0A 0x2F 0x2F 0x20 0x20 0x20 0x20 0x20 0x20 )
  ld: archive has no table of contents file '../lib/libfltk_jpeg.a' for architecture x86_64
  clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
  gmake[2]: *** [fltk/fluid/CMakeFiles/fluid.dir/build.make:488: fltk/bin/fluid.app/Contents/MacOS/fluid] Error 1
  gmake[1]: *** [CMakeFiles/Makefile2:407: fltk/fluid/CMakeFiles/fluid.dir/all] Error 2
  gmake: *** [Makefile:136: all] Error 2
  thread 'main' panicked at '
  command did not execute successfully, got: exit status: 2

**Comment by MoAlyousef** [comment]
I think you need to set CC and CXX to the correct compiler. Currently they point to the linker. 
Maybe:
CC="o64-clang" CXX="o64-clang++"

**Reply by cstkingkey** [reply]
the name may be confusing. the actual cc is x86_64-apple-darwin18-clang. using o64-clang-ld instead as clang will invoke /usr/bin/ld resulting errors.

**Reply by cstkingkey** [reply]
please make it able to set CMAKE_RANLIB  CMAKE_AR CMAKE_SYSTEM_NAME by env vars.

Thanks for your time.

**Comment by cstkingkey** [comment]
setting CMAKE_RANLIB and CMAKE_AR solve the problem.

**Reply by MoAlyousef** [reply]
Although undocumented, the current build supports passing the path to a cmake toolchain file via the CFLTK_TOOLCHAIN env variable. 

You would create a my_toolchain.cmake file with the cmake variables you need:
```cmake
set(CMAKE_AR "path/to/ar")
set(CMAKE_RANLIB "path/to/ranlib")
```
Then the build can be invoked using:
`CFLTK_TOOLCHAIN=my_toolchain.cmake cargo build --target x86_64-apple-darwin`

I’ve made a change in the build script to detect changes to that env variable.
